File: 백준/Gold/16929. Two Dots/Two Dots.cc
#include<iostream>
#include<string>

using namespace std;

int N, M;
bool sol;
char map[51][51] = { 0, };
int visit[51][51] = { 0, };

int dy[4] = { 0, 1, 0, -1 };
int dx[4] = { 1, 0, -1, 0 };

int start_y;
int start_x;

void dfs(int y, int x, int count) {

	visit[y][x] = 1;

	for (int i = 0; i < 4; i++) {
		int ry = y + dy[i];
		int rx = x + dx[i];

		if (ry >= 0 && ry < N && rx >= 0 && rx < M && map[y][x] == map[ry][rx] && visit[ry][rx] == 0) {
			dfs(ry, rx, count + 1);
		}
		else if (start_y == ry && start_x == rx && count >= 4) {
			sol = true;
			return;
		}
	}

}

void init_visit() {
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++) {
			visit[i][j] = 0;
		}
	}
}

int main() {

	cin >> N >> M;

	for (int i = 0; i < N; i++) {
		string str;

		cin >> str;
		for (int j = 0; j < M; j++) {
			map[i][j] = str[j];
		}
	}

	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++) {

			start_y = i;
			start_x = j;
			dfs(i, j, 1);

			if (sol == true) {
				cout << "Yes";
				return 0;
			}

			init_visit();
		}
	}

	cout << "No";
	return 0;
}

File: 백준/Gold/16929. Two Dots/Two Dots.java
import java.io.*;

public class Main {

    public static int N;
    public static int M;
    public static char[][] map;
    public static int[][] check;
    public static int[] m_r = {0,1,0,-1};
    public static int[] m_c = {1,0,-1,0};
    public static String sol = "No";
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        String[] str = br.readLine().split(" ");
        N = Integer.parseInt(str[0]);
        M = Integer.parseInt(str[1]);
        map = new char[N][M];
        check = new int[N][M];

        // map 세팅
        for (int i = 0; i < N; i++) {
            String str2 = br.readLine();
            for (int j = 0; j < M; j++) {
                map[i][j] = str2.charAt(j);
            }
        }

        for (int i=0; i < N; i++) {
            for (int j=0; j < M; j++) {
                if (check[i][j] == 0) {
                    dfs(i,j,i, j, 0);
                }
            }
            if (sol.equals("Yes")) {
                break;
            }
        }

        bw.write(sol+"\n");
        bw.flush();
        bw.close();
        br.close();
    }

    public static void dfs(int s_row, int s_col, int row, int col, int count) {

        check[row][col] = 1;
        for (int i=0; i<4; i++) {
            // 이동위치가 map 범위 내
            if (0<=row+m_r[i] && row+m_r[i]<N && 0<=col+m_c[i] && col+m_c[i]<M) {
                // 같은 문자이고
                if (map[row+m_r[i]][col+m_c[i]] == map[row][col]) {
                    if (count>2 && s_row==row+m_r[i] && s_col==col+m_c[i]) {
                        sol = "Yes";
                        return;
                    }
                    // 거쳐가지 않은 지점일때
                    if (check[row+m_r[i]][col+m_c[i]] == 0) {
                        dfs(s_row, s_col, row + m_r[i], col + m_c[i], count + 1);
                        check[row + m_r[i]][col + m_c[i]] = 0;
                    }
                }
            }
        }
    }
}